### PR TITLE
catalog,sqlbase,*: generalize descriptors, pass around wrapper structs

### PR DESCRIPTION
**Pull request by ajwerner**
This PR works towards enabling unified leasing of descriptors. I'm somewhat dissatisfied at the state it ends up leaving things but it is forward progress and I can't keep holding on to this.

This PR began its life with an intention of creating a unified `DescriptorMeta` proto message which would contain shared fields and live on the `Descriptor`. That change would necessitate not passing around the raw members of `Descriptor.Union` as they would no longer retain the necessary fields. Ultimately, after this was mostly done, it was determined that that change was not worth the migration pain. Nevertheless, a good bit of work was done to plumb through the new higher level abstractions and that work it valuable.

The biggest issue with reviewing this change will be `sqlbase,*: replace DescriptorProto with DescriptorInterface`, that commit does more than it should in a single commit but the time it is going to take to rip it apart doesn't seem obviously worth it given the overhead of keeping this change in sync.

I want to re-iterate that this change represents a stepping stone and not an end-state. The plan is to proceed from here by ripping apart sqlbase into subpackages such that the implementations of each of the wrapper types lives in their own packages. Then the plan will be to replace the use of all concrete fields in the descriptor types with interfaces. At that point the `Mutable` vs `Immutable` distinction can be rephrased as an interface distinction.

Doing the above should also aid in a number of other ways.

### CODE REVIEW

File: pkg/sql/sqlbase/structured.proto [hunk 1]
```diff
@@ -664,6 +737,27 @@ message TableDescriptor {
   // hlc timestamp to ensure that this field is set properly when extracted from
   // a Descriptor.
   optional util.hlc.Timestamp modification_time = 7 [(gogoproto.nullable) = false];
+  // A list of draining names. The draining name entries are drained from
+  // the cluster wide name caches by incrementing the version for this
+  // descriptor and ensuring that there are no leases on prior
+  // versions of the descriptor. This field is then cleared and the version
+  // of the descriptor incremented.
+  repeated NameInfo draining_names = 21 [(gogoproto.nullable) = false];
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.proto:745]: Is it safe/backwards compatible to move a message out of another message like this?

File: pkg/sql/sqlbase/structured.proto [hunk 2]
```diff
@@ -1019,9 +1026,16 @@ message DatabaseDescriptor {
   // Needed for the descriptorProto interface.
   option (gogoproto.goproto_getters) = true;
 
+  // Shared descriptor fields. See the discussion at the top of TableDescriptor.
+
   optional string name = 1 [(gogoproto.nullable) = false];
   optional uint32 id = 2 [(gogoproto.nullable) = false,
       (gogoproto.customname) = "ID", (gogoproto.casttype) = "ID"];
+  // Last modification time of the descriptor.
+  optional util.hlc.Timestamp modification_time = 4 [(gogoproto.nullable) = false];
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.proto:1035]: I thought you were going to pull modtime, version and draining names into one bigger message for the new descriptors

File: pkg/sql/sqlbase/structured.go [hunk 3]
```diff
@@ -3699,22 +3699,56 @@ func (desc *Descriptor) GetName() string {
...
-// which unmarshal Descriptor structs properly set the ModificationTime on
-// tables based on the MVCC timestamp at which the descriptor was read.
-//
-// A linter should ensure that GetTable() is not called.
-func (desc *Descriptor) Table(ts hlc.Timestamp) *TableDescriptor {
-	t := desc.GetTable()
-	if t != nil {
-		t.maybeSetTimeFromMVCCTimestamp(ts)
+// GetVersion returns the Version of the descriptor.
+func (desc *Descriptor) GetVersion() DescriptorVersion {
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:3703]: Should these panic if the type doesn't match? We want to defend against people adding new descriptors without adding the cases here too.

File: pkg/sql/sqlbase/structured.go [hunk 4]
```diff
@@ -3728,53 +3762,70 @@ func (desc *Descriptor) Table(ts hlc.Timestamp) *TableDescriptor {
...
-		log.Fatalf(context.TODO(), "table descriptor for %q (%d.%d) is in the "+
-			"ADD state and was created with CREATE TABLE ... AS but does not have a "+
-			"CreateAsOfTime set", desc.Name, desc.ParentID, desc.ID)
+func (desc *Descriptor) MaybeSetModificationTimeFromMVCCTimestamp(
+	ctx context.Context, ts hlc.Timestamp,
+) {
+	switch t := desc.Union.(type) {
+	case *Descriptor_Table:
+		// CreateAsOfTime is used for CREATE TABLE ... AS ... and was introduced in
+		// v19.1. In general it is not critical to set except for tables in the ADD
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:3771]: nit - ADD ADD state

File: pkg/sql/sqlbase/schema_desc.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+// ImmutableSchemaDescriptor wraps a Schema descriptor and provides methods
+// on it.
+type ImmutableSchemaDescriptor struct {
+	SchemaDescriptor
+}
+
+// MutableSchemaDescriptor is a mutable reference to a SchemaDescriptor.
+//
+// TODO(ajwerner): Today this isn't actually ever mutated but rather exists for
```
Comment (rohany) [inline pkg/sql/sqlbase/schema_desc.go:30]: I don't think it's that bad, especially if we embed the immutable one within the mutable one.

**Comment by rohany** [comment]
Overall, I'm on board with the changes made in this PR

File: pkg/sql/sqlbase/structured.proto (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/sqlbase/structured.proto:745]: Yes, protos are defined as just their set of fields IIUC. The nesting is really just a naming thing for the generated code and for interacting with imports.

File: pkg/sql/sqlbase/structured.proto (hunk 2 above)
Comment (ajwerner) [reply pkg/sql/sqlbase/structured.proto:1035]: Discussed offline, copying the fields allows us to remove the migration. My preference evolved over time to preferring the messier protos and proto handling code in favor of the simpler go code given the existing structure.

File: pkg/sql/sqlbase/schema_desc.go (hunk 5 above)
Comment (ajwerner) [reply pkg/sql/sqlbase/schema_desc.go:30]: Moved to a `Note` to clarify the intention to keep changing the structure here.

File: pkg/sql/sqlbase/structured.go (hunk 3 above)
Comment (ajwerner) [reply pkg/sql/sqlbase/structured.go:3702]: Yeah, done.

File: pkg/sql/sqlbase/structured.go (hunk 4 above)
Comment (ajwerner) [reply pkg/sql/sqlbase/structured.go:3772]: Done.

**Review by thoszhang — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r=rohany,lucy-zhang

**Comment by rohany** [comment]
Can't believe that made it in on the first bors try

**Pull request merged**